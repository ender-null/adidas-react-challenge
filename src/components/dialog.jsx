import React, { useState } from "react";

export const Dialog = ({ round, ranking, onSubmit }) => {
  const [name, setName] = useState("");

  const handleInput = (event) => {
    setName(event.target.value);
  };

  const handleSubmit = () => {
    ranking.push({ name, round });
    localStorage.setItem("ranking", JSON.stringify(ranking));
    onSubmit();
  };

  return (
    <div className="overlay">
      <div className="dialog">
        <form onSubmit={handleSubmit}>
          <h2 className="header">Wrong tile!</h2>
          <input
            className="input"
            onChange={handleInput}
            placeholder="Write your name"
            autoFocus
          ></input>
          <button className="button" disabled={name.length === 0} type="submit">
            Confirm
          </button>
        </form>
      </div>
    </div>
  );
};
