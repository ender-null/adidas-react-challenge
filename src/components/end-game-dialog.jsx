import React, { useEffect, useState } from "react";

export const EndGameDialog = ({
  step,
  previousName,
  onSetName,
  onAddEntry,
  onCancel,
}) => {
  const [name, setName] = useState(previousName);

  const handleInput = (event) => {
    setName(event.target.value.trim());
  };

  useEffect(() => {
    onSetName(name);
  }, [name, onSetName]);

  return (
    <div className="overlay">
      <div className="dialog">
        <button
          className="close-button"
          id="end-game-dialog-submit"
          onClick={onCancel}
        >
          X
        </button>
        <form onSubmit={onAddEntry}>
          <h2 className="header">Wrong tile!</h2>
          {step > 0 && (
            <>
              <span>
                You got to step <strong>{step}</strong>
              </span>
              <input
                id="end-game-dialog-input"
                className="input"
                value={name}
                onChange={handleInput}
                placeholder="Write your name"
                autoFocus
              ></input>
            </>
          )}
          <button
            id="end-game-dialog-submit"
            className="button"
            disabled={(!name || name.length === 0) && step > 0}
            type="submit"
          >
            Accept
          </button>
        </form>
      </div>
    </div>
  );
};
