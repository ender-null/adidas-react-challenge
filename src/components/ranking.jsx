import React from "react";

export const Ranking = ({ ranking, onNewGame }) => {
  const orderedRanking = Object.entries(ranking);
  orderedRanking.sort(([, stepA], [, stepB]) => {
    return stepB - stepA;
  });

  return (
    <div className="ranking">
      <h1 className="header">Hall of fame</h1>
      <ol>
        {orderedRanking.map(([name, step], index) => {
          return (
            <li key={index}>
              <strong>{name}</strong> {step}
            </li>
          );
        })}
      </ol>
      <button className="button" onClick={onNewGame}>
        New game
      </button>
    </div>
  );
};
