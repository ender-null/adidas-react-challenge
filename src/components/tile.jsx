import React from "react";

export const Tile = ({ style, onClick, isDifferent }) => {
  return (
    <div
      className={isDifferent ? "tile different" : "tile"}
      style={style}
      onClick={onClick}
    ></div>
  );
};
