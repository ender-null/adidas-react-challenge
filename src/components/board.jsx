import React, { useEffect, useState } from "react";
import { Tile } from "./tile.jsx";

const BoardWidth = 320;

export const Board = ({ round, onSuccess, onFail }) => {
  const [tiles, setTiles] = useState(null);

  const getSize = () => round + 2;

  const createTile = (differentTileIndex, color) => (_, index) => {
    const isDifferent = index === differentTileIndex;
    const style = {
      background: `hsl(${color}, 60%, ${isDifferent ? 65 : 50}%)`,
      width: `calc(100% / ${getSize()} - 4px)`,
      height: `calc(100% / ${getSize()} - 4px)`,
    };

    return (
      <Tile
        key={index}
        style={style}
        onClick={isDifferent ? onSuccess : onFail}
      />
    );
  };

  function createTiles() {
    const size = getSize();
    const totalSize = size * size;
    const highlightTile = Math.floor(Math.random() * totalSize);
    const color = Math.floor(Math.random() * 256);
    const newTiles = Array(totalSize)
      .fill(null)
      .map(createTile(highlightTile, color));
    setTiles(newTiles);
  }

  useEffect(() => {
    createTiles();
  }, [round]);

  return (
    <div className="board">
      <div className="status">
        Round: <strong>{round}</strong>
      </div>
      <div
        className="tiles"
        style={{ width: `${BoardWidth}px`, height: `${BoardWidth}px` }}
      >
        {tiles}
      </div>
    </div>
  );
};
