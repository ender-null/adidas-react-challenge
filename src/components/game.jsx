import React, { useEffect, useState } from "react";
import { Board } from "./board.jsx";
import { EndGameDialog } from "./end-game-dialog.jsx";
import { Ranking } from "./ranking.jsx";

export const Game = () => {
  const [name, setName] = useState("");
  const [step, setStep] = useState(Number(localStorage.getItem("step")) || 0);
  const [ranking, setRanking] = useState(
    JSON.parse(localStorage.getItem("ranking")) || {}
  );
  const [showDialog, setShowDialog] = useState(
    localStorage.getItem("show_dialog")
  );

  const handleSuccess = () => {
    setStep(step + 1);
  };

  const handleFail = () => {
    setShowDialog(true);
  };

  const handleNewGame = () => {
    setStep(0);
  };

  const handleAddEntry = () => {
    if (step > 0) {
      setRanking({
        ...ranking,
        [name]: ranking[name] && ranking[name] > step ? ranking[name] : step,
      });
      setStep(-1);
    }
    setShowDialog(false);
  };

  const handleCancel = () => {
    if (step > 0 && ranking.length > 0) {
      setStep(-1);
    } else {
      setStep(0);
    }
    setShowDialog(false);
  };

  useEffect(() => {
    localStorage.setItem("step", step);
  }, [step]);

  useEffect(() => {
    localStorage.setItem("ranking", JSON.stringify(ranking));
  }, [ranking]);

  useEffect(() => {
    localStorage.setItem("show_dialog", showDialog);
  }, [showDialog]);

  if (step < 0) {
    return (
      <div className="game">
        <Ranking ranking={ranking} onNewGame={handleNewGame}></Ranking>
      </div>
    );
  }

  return (
    <div className="game">
      <Board step={step} onSuccess={handleSuccess} onFail={handleFail} />
      {showDialog && (
        <EndGameDialog
          onSetName={setName}
          previousName={name}
          step={step}
          onAddEntry={handleAddEntry}
          onCancel={handleCancel}
        ></EndGameDialog>
      )}
    </div>
  );
};
