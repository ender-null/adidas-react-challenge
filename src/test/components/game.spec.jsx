import { render } from "@testing-library/react";
import { Game } from "../../components/game";

describe("Game", () => {
  beforeEach(() => {
    jest.spyOn(global.Math, "random").mockReturnValue(0.2);
  });

  afterEach(() => {
    jest.spyOn(global.Math, "random").mockRestore();
  });

  test("should render the component", () => {
    const { asFragment } = render(<Game />);
    expect(asFragment()).toMatchSnapshot();
  });
});
