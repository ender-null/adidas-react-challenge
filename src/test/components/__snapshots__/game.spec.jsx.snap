// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Game should render the component 1`] = `
<DocumentFragment>
  <div
    class="game"
  >
    <div
      class="board"
    >
      <div
        class="header"
      >
        Step: 
        <strong>
          0
        </strong>
      </div>
      <div
        class="tiles"
        style="width: 320px; height: 320px;"
      >
        <div
          class="tile different"
          style="background: rgb(112, 112, 112); width: calc(100% / 2 - 4px); height: calc(100% / 2 - 4px);"
        />
        <div
          class="tile"
          style="background: rgb(51, 51, 51); width: calc(100% / 2 - 4px); height: calc(100% / 2 - 4px);"
        />
        <div
          class="tile"
          style="background: rgb(51, 51, 51); width: calc(100% / 2 - 4px); height: calc(100% / 2 - 4px);"
        />
        <div
          class="tile"
          style="background: rgb(51, 51, 51); width: calc(100% / 2 - 4px); height: calc(100% / 2 - 4px);"
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;
